src/aspire/denoising/denoiser_cov2d.py

```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
-    def __init__(self, src, basis, var_noise, ctf_info):
+    def __init__(self, src, basis, var_noise, ctf_info, star_file):
         """
         Initialize an object for denoising 2D images using Cov2D method
 
         :param src: The source object of 2D images with metadata
         :param basis: The basis method to expand 2D images
         :param var_noise: The estimated variance of noise
         :param ctf_info: Whether the CTF filters are included
+        :param star_file: the star file path and name to output
         """
         super().__init__(src)
-        self.imgs_noise = self.src.images(start=0, num=np.inf).asnumpy()
         self.var_noise = var_noise
         if not isinstance(basis, FFBBasis2D):
             raise NotImplementedError('Currently only fast FB method is supported')
-        self.coeffs_noise = None
-        self.coeffs_estim = None
-        self.ctf_idx = None
-        self.ctf_fb = None
         self.basis = basis
-        self.imgs_estim = None
+        self.ctf_info = ctf_info
+        self.star_file = star_file
+        self.starfile_path = os.path.dirname(star_file)
+        self.starfile_name = os.path.basename(star_file)
 
+        self.ctf_idx = None
+        self.ctf_fb = None
         # Assign the CTF information and index for each image
         if ctf_info and not (src.filters is None):
             logger.info(f'Represent CTF filters in FB basis')
-            uniq_ctf_filters = list(set(src.filters))
-            # Create the indices of CTF filters from all images
-            self.ctf_idx = np.array([uniq_ctf_filters.index(f) for f in src.filters])
-            # Evaluate CTFs in the FFB basis
-            self.ctf_fb = [f.fb_mat(basis) for f in uniq_ctf_filters]
         else:
             logger.info(f'CTF filters are not included in Cov2D denoising')
             # set all CTF filters to an identity filter
             self.ctf_idx = np.zeros(self.nimg, dtype=int)
             self.ctf_fb = [blk_diag_eye(blk_diag_partition(RadialCTFFilter().fb_mat(basis)))]
 
-    def denoise(self, covar_opt=None):
+        self.metadata = src._metadata.copy(deep=True)
+        self._write_star_file()
+        self.mrcs_out = self._get_mrcs_out()
+
+    def denoise(self, covar_opt=None, overwrite=False):
         """
         Denoiser 2D images using Cov2D method
 
-        :param basis: The basis method to expand 2D images
         :param covar_opt: The option list for building Cov2D matrix
+        :param overwrite: The option to owerwrite mrcs files for denoised images or not
         """
+        src = self.src
 
-        self.coeffs_noise = self.basis.evaluate_t(self.imgs_noise)
-
-        cov2d = RotCov2D(self.basis)
+        # Initialize the rotationally invariant covariance matrix of 2D images
+        # A fixed batch size is used to go through each image
+        cov2d = BatchedRotCov2D(self.src, self.basis, batch_size=512)
+        if not self.ctf_info:
+            cov2d.ctf_idx = self.ctf_idx
```

Comment:
Again, this shouldn't be needed. Better to handle it directly in `BatchedRotCov2D` (or even in `ImageSource`).

Reply:
As above, it is moved to `BatchedCov2D` in new commit f83b805.